pkg/broker/receiver.go

```diff
@@ -0,0 +1,141 @@
+ */
+
+package broker
+
+import (
+	"context"
+	"errors"
+
+	eventingv1alpha1 "github.com/knative/eventing/pkg/apis/eventing/v1alpha1"
+	"github.com/knative/eventing/pkg/provisioners"
+	"go.uber.org/zap"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+)
+
+// Receiver parses Cloud Events, determines if they pass a filter, and sends them to a subscriber.
+type Receiver struct {
+	logger *zap.Logger
+	client client.Client
+
+	dispatcher provisioners.Dispatcher
+}
+
+// New creates a new Receiver and its associated MessageReceiver. The caller is responsible for
+// Start()ing the returned MessageReceiver.
+func New(logger *zap.Logger, client client.Client) (*Receiver, manager.Runnable) {
+	r := &Receiver{
+		logger:     logger,
+		client:     client,
+		dispatcher: provisioners.NewMessageDispatcher(logger.Sugar()),
+	}
+	return r, r.newMessageReceiver()
+}
+
+func (r *Receiver) newMessageReceiver() *provisioners.MessageReceiver {
+	if err := r.initClient(); err != nil {
+		r.logger.Warn("Failed to initialize client", zap.Error(err))
+	}
+	return provisioners.NewMessageReceiver(r.sendEvent, r.logger.Sugar())
+}
+
+// sendEvent sends an event to a subscriber if the trigger filter passes.
+func (r *Receiver) sendEvent(trigger provisioners.ChannelReference, message *provisioners.Message) error {
+	r.logger.Debug("Received message", zap.Any("triggerRef", trigger))
+	ctx := context.Background()
+
+	t, err := r.getTrigger(ctx, trigger)
+	if err != nil {
+		r.logger.Info("Unable to get the Trigger", zap.Error(err), zap.Any("triggerRef", trigger))
+		return err
+	}
+
+	subscriberURI := t.Status.SubscriberURI
+	if subscriberURI == "" {
+		r.logger.Error("Unable to read subscriberURI")
+		return errors.New("unable to read subscriberURI")
+	}
+
+	if !r.shouldSendMessage(&t.Spec, message) {
+		r.logger.Debug("Message did not pass filter", zap.Any("triggerRef", trigger))
+		return nil
+	}
+
+	err = r.dispatcher.DispatchMessage(message, subscriberURI, "", provisioners.DispatchDefaults{})
+	if err != nil {
+		r.logger.Info("Failed to dispatch message", zap.Error(err), zap.Any("triggerRef", trigger))
+		return err
+	}
+	r.logger.Debug("Successfully sent message", zap.Any("triggerRef", trigger))
+	return nil
+}
+
+// Initialize the client. Mainly intended to load stuff in its cache.
+func (r *Receiver) initClient() error {
+	// We list triggers so that we can load the client's cache. Otherwise, on receiving an event, it
+	// may not find the trigger and would return an error.
+	opts := &client.ListOptions{
+		// Set Raw because if we need to get more than one page, then we will put the continue token
+		// into opts.Raw.Continue.
+		Raw: &metav1.ListOptions{},
+	}
+	for {
+		tl := &eventingv1alpha1.TriggerList{}
+		if err := r.client.List(context.TODO(), opts, tl); err != nil {
+			return err
+		}
+		if tl.Continue != "" {
```

Comment:
I don't think this will work. 2  reasons:

1. It seems that at this point when r.client.List is called. It will just create the informer/indexer internally but will not run and sync cache. Which means tl will be nil. The cache is synced when manager starts. 
2. Even if it returns something, tl will always be reset and not appended. Moreover, controller-fwk client does all the list and gets from cache and the caches are populated async. I don't think it will honor continue, as the entire list is in-memory and is returned.

Please enable diagnostic logging and verify this.

Reply:
I can take care of this tomorrow. We don't care about the continue, it was a copy paste from other places that are listing things as this. 
I had to add this ugly method in order to make the e2e test work. Otherwise we were losing few events. I can try removing this, and see if the e2e passes.

Reply:
I just tested removing this initClient but the e2e test doesn't work.
We may need to update the comment of the method to say that we need to create the informer/indexer in order not to drop messages, instead of populating the cache.
Removing the paging does work so I think we should remove it. The code would look like this:

```
	opts := &client.ListOptions{}
	tl := &eventingv1alpha1.TriggerList{}
	if err := r.client.List(context.TODO(), opts, tl); err != nil {
		return err
	}
	return nil
```
Thanks @akashrv for the comment
@Harwayne what do you think?

Reply:
Done.